language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala

```diff
@@ -241,22 +242,21 @@ object CodeGen {
     // or Scala 2.13
     val (deletedRecords, newVariants) =
       variants.traverseU {
-        case vsdt @ ScopedDataType(Identifier(packageId, qualName), vTypeVars, _) =>
+        case ScopedDataType(ident @ Identifier(packageId, qualName), vTypeVars, Variant(fields)) =>
           val typeVarDelegate = Util simplyDelegates vTypeVars
-          vsdt.traverseU {
-            _.traverseU {
-              case (vn, vt) =>
-                val syntheticRecord = Identifier(
-                  packageId,
-                  qualName copy (name =
-                    DottedName.assertFromSegments(qualName.name.segments.slowSnoc(vn).toSeq)))
-                val key = (syntheticRecord, vTypeVars.toList)
-                typeVarDelegate(vt)
-                  .filter((_: Identifier) == syntheticRecord)
-                  .flatMap(_ => recordMap get key)
-                  .cata(nr => (Set(key), (vn, -\/(nr.fields.toList))), (noDeletion, (vn, \/-(vt))))
-            }
+          val (deleted, sdt) = fields.traverseU {
+            case (vn, vt) =>
+              val syntheticRecord = Identifier(
+                packageId,
+                qualName copy (name =
+                  DottedName.assertFromSegments(qualName.name.segments.slowSnoc(vn).toSeq)))
+              val key = (syntheticRecord, vTypeVars.toList)
+              typeVarDelegate(vt)
+                .filter((_: Identifier) == syntheticRecord)
+                .flatMap(_ => recordMap get key)
+                .cata(nr => (Set(key), (vn, -\/(nr.fields.toList))), (noDeletion, (vn, \/-(vt))))
           }
+          (deleted, ScopedDataType(ident, vTypeVars, Variant(sdt)))
```

Comment:
I’d rather keep the traverse layers; this reversion used to be how it was but it caused confusing flexibility problems with the original introduction of `ScopedDataType`. One of the many times I've been punished for ignoring abstractions I'm aware of in favor of "intuitive" inlined code. I would be ok with this if I believed that the `interface` library was not going to have any design changes in the future :)

Reply:
Eh, I guess the choice of taking the `(String,)` out of the tparam makes that impossible, never mind